File: PadraoGof/Criacionais/FactoryMethod/factory.py
#class base veiculo
class Veiculo:
    def __init__(self, modelo):
        self.modelo = modelo

    def mostrarDetalhes(self):
        print(f'Modelo: {self.modelo}')

#subclasse de veiculo
class Carro(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

class Moto(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

class Navio(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

#fabrica abstrata de veiculos
class FabricaVeiculo:
    def criarVeiculo(self, modelo):
        raise NotImplementedError ("o metodo deve ser implementado pela subclasse")

#fabrica concreta de carro
class FabricaCarro (FabricaVeiculo):
    def criarVeiculo(self, modelo):
        return Carro(modelo)

# fabrica concreta de motos
class FabricaMoto(FabricaVeiculo):
    def criarVeiculo(self, modelo):
        return Moto (modelo)

#fabrica concreta de navios
class FabricaNavio(FabricaVeiculo):
    def criarVeiculo(self, modelo):
        return Navio (modelo)

#uso padrao..........
fabricaMoto = FabricaMoto()
fabricaCarro = FabricaCarro()
fabricaNavio = FabricaNavio()

transporte1 = fabricaCarro.criarVeiculo('sedan') 
transporte2 = fabricaCarro.criarVeiculo('rete')
transporte3 = fabricaMoto.criarVeiculo('bigtrail')
transporte4 = fabricaMoto.criarVeiculo('esportiva')
transporte5 = fabricaNavio.criarVeiculo('cargueiro')
transporte6 = fabricaNavio.criarVeiculo('container')
 

transporte1.mostrarDetalhes()
transporte2.mostrarDetalhes()
transporte3.mostrarDetalhes()
transporte4.mostrarDetalhes()
transporte5.mostrarDetalhes()
transporte6.mostrarDetalhes()

File: PadraoGof/Criacionais/FactoryMethod/factory.js
//classe base de veiculos
class Veiculos{
    constructor (modelo){
        this.modelo = modelo;
    }
    mostrarDetalhe(){
        console.log(`Modelo: ${this.modelo}`);
    }
} 
//subclasse de veiculos 
class Carros extends Veiculos{
    constructor (modelo){
    super (modelo);
    }
}
class Motos extends Veiculos{
    constructor (modelo){
    super (modelo);
    }
}
class Navios extends Veiculos{
    constructor (modelo){
        super (modelo);
    }
}
//fabrica abstrata de veiculos

class FabricaVeiculos{
    criarVeiculo(modelo){
        throw new Error (`o metodo deve ser implementado pela subclasse`)
    }
}
//fabrica concreta de carro 
class FabricaCarros extends FabricaVeiculos{
    criarVeiculo(modelo){
        return new Carros(modelo);
    }
}
//fabrica concreta de motos 
class FabricaMotos extends FabricaVeiculos{
    criarVeiculo(modelo){
        return new Motos (modelo);
    }
}
//fabrica concretas de navio
class FabricaNavios extends FabricaVeiculos{
    criarVeiculo(modelo){
    return new Navios (modelo);
    }
}
// uso padrao.........
const fabricaMotos = new FabricaMotos();
const fabricaCarros = new FabricaCarros();
const fabricaNavio = new FabricaNavios();

const transporte1 = fabricaCarros.criarVeiculo('sedan'); 
const transporte2 = fabricaCarros.criarVeiculo('rete');
const transporte3 = fabricaMotos.criarVeiculo('bigtrail');
const transporte4 = fabricaMotos.criarVeiculo('esportiva');
const transporte5 = fabricaNavio.criarVeiculo('cargueiro');
const transporte6 = fabricaNavio.criarVeiculo('container');
 
transporte1.mostrarDetalhe();
transporte2.mostrarDetalhe();
transporte3.mostrarDetalhe();
transporte4.mostrarDetalhe();
transporte5.mostrarDetalhe();
transporte6.mostrarDetalhe();
